fix(219): corrigir falha na aceitação de propostas

anthonysfarias opened the pull request:
Correção: Falha na aceitação de propostas

### Problema Identificado
O sistema estava falhando ao aceitar propostas devido a dois problemas principais:

1. **Erro de referência de objetos**: O código assumia que `proposal.bid` sempre seria um objeto com propriedade `.id`, mas dependendo do contexto de população do MongoDB/Mongoose, pode ser uma string ou objeto com `_id`
2. **Falta de validação de status**: Não havia verificação se lotes estavam em análise antes de aceitar/recusar propostas

### Soluções Implementadas

#### Tratamento robusto de referências
- Implementado verificação de tipo para `proposal.bid` e `proposal.proposedBy.supplier`
- Suporte para diferentes formatos: string, objeto com `_id` ou `id`
- Evita erros "Cannot read property 'id' of undefined"

#### Nova validação de regra de negócio
- Adicionada verificação de status `AllotmentStatusEnum.emAnalise` nos lotes
- Impede aceitar/recusar propostas quando lotes estão em análise
- Retorna erro claro: "Não é possível aceitar propostas enquanto o lote está em análise"

### Arquivos Alterados
- [backend/src/modules/SOL/services/proposal.service.ts](cci:7://file:///c:/Users/trabalho/apps/staging/219-investigar-falha-aceitacao-propostas/backend/src/modules/SOL/services/proposal.service.ts:0:0-0:0)
- Adicionadas validações nos métodos de aceitar e recusar propostas

anthonysfarias committed:
fix(219): corrigir falha na aceitação de propostas

anthonysfarias committed:
tests: created unit tests for proposal.service.spec.ts

anthonysfarias committed:
tests: created unit tests for proposal.service.spec.ts

juanmeleiro commented:
Merge fica pendente esperando o PR correspondente no frontend.

gabrielmussolini984 commented on src/modules/SOL/services/proposal.service.ts:
```diff
@@ -433,7 +433,11 @@ export class ProposalService {
     const proposal = await this._proposalRepository.getById(proposalId);
 
-    const list = await this._proposalRepository.listByBid(proposal.bid.id);
+    const list = await this._proposalRepository.listByBid(
+      typeof proposal.bid === "string"
+        ? proposal.bid
+        : proposal.bid._id || proposal.bid.id,
+    );
```
Isso aqui não faz sentido, tem alguma inconsistencia no banco?

gabrielmussolini984 replied:
O ideal é resolver a inconsistência

anthonysfarias replied:
Limpei código repetitivo no sistema de propostas que estava causando problemas de manutenção.

O código tinha várias verificações iguais espalhadas:
- Verificava se proposal.bid era string ou objeto
- Isso acontecia em 5 lugares diferentes
- Código difícil de manter e propenso a erros

**Criei funções auxiliares**
- extractBidId() - pega ID do bid
- extractSupplierId() - pega ID do fornecedor
- extractEntityId() - função genérica

**Atualizei o ProposalService**
- Removi as 5 verificações repetitivas
- Usei as funções auxiliares no lugar

**Padronizei o Repository**
- Todos os métodos agora retornam dados consistentes
- Campo bid sempre vem populado do banco


## Arquivos que alterei

- [proposal.service.ts](cci:7://file:///c:/Users/trabalho/apps/staging/219-investigar-falha-aceitacao-propostas/backend/src/modules/SOL/services/proposal.service.ts:0:0-0:0) - removi verificações repetitivas
- [proposal.repository.ts](cci:7://file:///c:/Users/trabalho/apps/staging/219-investigar-falha-aceitacao-propostas/backend/src/modules/SOL/repositories/proposal.repository.ts:0:0-0:0) - padronizei população de dados
- [entity-id-extractor.helper.ts](cci:7://file:///c:/Users/trabalho/apps/staging/219-investigar-falha-aceitacao-propostas/backend/src/modules/SOL/utils/entity-id-extractor.helper.ts:0:0-0:0) - criei novas funções auxiliares

gabrielmussolini984 commented on src/modules/SOL/services/proposal.service.ts:
```diff
@@ -494,6 +498,18 @@ export class ProposalService {
       const proposal = await this._proposalRepository.getById(proposalId);
+
+      // Validação: impedir recusa se algum lote estiver em análise
+      if (
+        proposal.allotment &&
+        proposal.allotment.some(
+          (a) => a.status === AllotmentStatusEnum.emAnalise,
+        )
```
```suggestion
const hasAllotmentInAnalysis= proposal.allotment?.some(
          (a) => a.status === AllotmentStatusEnum.emAnalise
        )
        
      if (hasAllotmentUnderAnalysis )
         
```

gabrielmussolini984 replied:
Jogando para uma const fica mais claro o some.
Se colocar o ? não precisa validar proposal.allotment

anthonysfarias replied:
Refatorei a verificação de allotment conforme sua sugestão

gabrielmussolini984 commented on src/modules/SOL/services/proposal.service.ts:
```diff
@@ -536,28 +552,56 @@ export class ProposalService {
+        proposal.allotment &&
+        proposal.allotment.some(
+          (a) => a.status === AllotmentStatusEnum.emAnalise,
+        )
```
Mesma coisa que o comentário acima

gabrielmussolini984 commented on src/modules/SOL/services/proposal.service.ts:
```diff
@@ -536,28 +552,56 @@ export class ProposalService {
       let contractDto: ContractRegisterDto = {
         contract_number: "1",
-        bid_number: proposal.bid.id,
+        bid_number:
+          typeof proposal.bid === "string"
+            ? proposal.bid
+            : proposal.bid._id || proposal.bid.id,
```
Iconsistência no banco?

gabrielmussolini984 commented on src/modules/SOL/services/proposal.service.ts:
```diff
@@ -536,28 +552,56 @@ export class ProposalService {
         proposal_id: [proposal],
         association_id: bid.id,
-        supplier_id: proposal.proposedBy.supplier.id,
+        supplier_id:
+          typeof proposal.proposedBy.supplier === "string"
+            ? proposal.proposedBy.supplier
+            : proposal.proposedBy.supplier._id ||
+              proposal.proposedBy.supplier.id,
```
mesma coisa

gabrielmussolini984 commented on src/modules/SOL/services/proposal.service.ts:
```diff
@@ -494,6 +498,18 @@ export class ProposalService {
+        )
+      ) {
+        throw new BadRequestException(
+          "Não é possível recusar propostas enquanto o lote está em análise.",
```
se possível, começar a passar os throw para um dict, para ficar fácil retornar os errors conforme a location

anthonysfarias replied:
Feito!

anthonysfarias replied:
Ainda vou ajustar as mensagens para o padrão i18n, mas por enquanto, implementei conforme a revisão

gabrielmussolini984 reviewed (commented):
Sempre que alterar alguma função, tente refatorar para deixar mais legivel e ajustar tipagens.

Faça primeiro os casos de testes validando o comportamento atual, pode explorar um pouco do it.each() para criar variações para o mesmo teste.

Fazendo isso, fica bem mais simples de fazer refactor e aumenta a cobertura dos testes.

anthonysfarias committed:
refactor(219): eliminar code smell em ProposalService

anthonysfarias committed:
refactor(219): improve allotment verification readability

anthonysfarias committed:
refactor(219): improve allotment verification readability

anthonysfarias committed:
refactor(219): improve allotment verification readability

anthonysfarias committed:
refactor(219): improve allotment verification readability

anthonysfarias committed:
refactor(219): improve allotment verification readability

anthonysfarias committed:
refactor(219): improve allotment verification readability

anthonysfarias committed:
refactor(219): improve allotment verification readability

anthonysfarias committed:
refactor(219): simplify error handling in proposal-accepted component

gabrielmussolini984 commented on src/modules/SOL/controllers/proposal.controller.ts:
```diff
@@ -445,8 +445,6 @@ export class ProposalController {
 
       return new ResponseDto(true, response, null);
     } catch (error) {
-      this.logger.error(error.message);
-
```
Pode deixar o logger no backend

gabrielmussolini984 committed:
Criação de um novo handler de erro

anthonysfarias committed:
feat: add structured error handling system

- Add BackendErrors enum and StructuredErrorHelper
- Update proposal service/controller for structured errors
- Enable automatic frontend error translation

anthonysfarias commented:
O código ainda está em manutenção!

anthonysfarias committed:
feat: implement structured error handling system

anthonysfarias committed:
feat: implement structured error handling system

anthonysfarias commented:
@gabrielmussolini984 Aguardando revisão
